File: pe/10.js
function processData(input) {
var max_n = 1000000
var sieve = []
var primes = []

for (var i = 0; i < 2    ; i++) sieve.push(false)
for (var i = 2; i < max_n; i++) sieve.push(true)

for (var i = 2; i < max_n; i++) {
    if (sieve[i]) {
        primes.push(i)
        for (var j = 2*i; j < max_n; j += i) {
            sieve[j] = false
        }
    }
}

function sum(lim) {
    var p = 0
    for (var i = 0; primes[i] <= lim; i++)
        p += primes[i]
    return p
}

var lines = input.split("\n")
var t = parseInt(lines[0])

for (var i = 1; i <= t; i++)
    console.log(sum(parseInt(lines[i])))

}

process.stdin.resume();
process.stdin.setEncoding("ascii");
_input = "";
process.stdin.on("data", function (input) {
    _input += input;
});

process.stdin.on("end", function () {
   processData(_input);
});

File: pe/10.hs
-- Too slow ... won't pass last test case

import Control.Monad.ST
import Control.Monad
import Data.STRef
import Data.Array.ST

max_n :: Int
max_n = 1000000

sieve :: UArray Int Bool
sieve = runSTUArray $
    newArray (2, max_n) True >>= \sieve ->
    forM_ [2 .. max_n] (\p ->
        readArray sieve p >>= \isPrime ->
        when isPrime
            (forM_ [p*2, p*3 .. max_n] (\k ->
                writeArray sieve k False))) >>
    return sieve

primes :: [Int]
primes = [i | (i,isPrime) <- assocs sieve, isPrime]

main :: IO ()
main =
    let toInt = read :: String -> Int in
    getLine >>= \t_str -> let t = toInt t_str in
    forM_ [1 .. t] (\_ ->
        getLine >>= \n_str -> let n = toInt n_str in
        print $ sum $ takeWhile (<= n) primes)

